gen/yarpc.go

```diff
@@ -0,0 +1,329 @@
+			<$Helper := printf "%s.%sHelper" $servicePackage (goCase .Name)>
+
+			<$req := newVar "req">
+			<$body := newVar "body">
+
+			func (h handler) <goCase .Name>(
+				<$req> *<$thrift>.Request,
+				<$body> <$wire>.Value,
+			) (<$wire>.Value, *<$thrift>.Response, error) {
+
+				<$args := newVar "args">
+				var <$args> <$Args>
+				if err := <$args>.FromWire(<$body>); err != nil {
+					return <$wire>.Value{}, nil, err
+				}
+
+				<$res := newVar "res">
+				<$succ := newVar "success">
+				<if .ResultSpec.ReturnType>
+					<$succ>,
+				<end>
+				<$res>, err := h.impl.<goCase .Name>(
+					<$req>,
+					<range .ArgsSpec><$args>.<goCase .Name>,<end>
+				)
+
+				<$result := newVar "result">
+				<$result>, err := <$Helper>.WrapResponse(
+					<if .ResultSpec.ReturnType>
+						<$succ>,
+					<end>
+					err)
+				return <$result>.ToWire(), <$res>, err
+			}
+		<end>
+		`,
+		s)
+
+	if err != nil {
+		return nil, err
+	}
+
+	buff := new(bytes.Buffer)
+	if err := yg.g.Write(buff, token.NewFileSet()); err != nil {
+		return nil, fmt.Errorf(
+			"failed to write YARPC server for service %q: %v", s.Name, err)
+	}
+
+	return buff, nil
+}
+
+func (yg yarpcGenerator) client(s *compile.ServiceSpec) (*bytes.Buffer, error) {
+	if err := yg.iface(s, false); err != nil {
+		return nil, err
+	}
+
+	err := yg.g.DeclareFromTemplate(
+		`
+		<$transport := import "github.com/yarpc/yarpc-go/transport">
+		<$thrift := import "github.com/yarpc/yarpc-go/encoding/thrift">
+		<$protocol := import "github.com/thriftrw/thriftrw-go/protocol">
+
+		func New(c <$transport>.Channel) Interface {
+			return client{
+				c: <$thrift>.New(<$thrift>.Config{
+					Service: "<.Name>",
+					Channel: c,
+					Protocol: <$protocol>.Binary,
+				}),
+			}
+		}
+
+		type client struct{c <$thrift>.Client}
+
+		<$service := .>
+		<range .Functions>
+			<$servicePackage := servicePackage $service>
+			<$Result := printf "%s.%sResult" $servicePackage (goCase .Name)>
+			<$Helper := printf "%s.%sHelper" $servicePackage (goCase .Name)>
+
+			<$params := newNamespace>
+			func (c client) <goCase .Name>(
+				<$params.NewName "req"> *<$thrift>.Request,
+				<range .ArgsSpec>
+					<if .Required>
+						<$params.NewName .Name> <typeReference .Type>,
+					<else>
+						<$params.NewName .Name> <typeReferencePtr .Type>,
+					<end>
+				<end>
+			) (
+				<if .ResultSpec.ReturnType>
+					<typeReference .ResultSpec.ReturnType>,
+				<end>
+				*<$thrift>.Response,
+				 error,
+			 ) {
+				<$req := $params.Rotate "req">
+				<$args := $params.NewName "args">
+				<$args> := <$servicePackage>.<goCase .Name>Helper.Args(
+					<range .ArgsSpec><$params.Rotate .Name>,<end>)
+
+				<$body := $params.NewName "body">
+				<$res := $params.NewName "res">
+				<$body>, <$res>, err := c.c.Call("<.Name>", <$req>, <$args>.ToWire())
+				if err != nil {
+					<if .ResultSpec.ReturnType>
+						return <zeroValue .ResultSpec.ReturnType>, <$res>, err
```

Comment:
instead of having a zeroValue, there's approaches to do it with less code (and probably less likely to have bugs):

- use named return value, and return that here
- create a temporary var,
```go
var emptyRet <.ResultSpec.ReturnType>
return emptyRet, <$res>, err
```

What do you think?